File: src/main/java/com/telegram/folobot/service/MessageService.java
package com.telegram.folobot.service;

import com.telegram.folobot.FoloBot;
import lombok.RequiredArgsConstructor;
import lombok.Setter;
import lombok.experimental.Accessors;
import org.springframework.stereotype.Component;
import org.telegram.telegrambots.meta.api.methods.ActionType;
import org.telegram.telegrambots.meta.api.methods.ForwardMessage;
import org.telegram.telegrambots.meta.api.methods.ParseMode;
import org.telegram.telegrambots.meta.api.methods.send.SendChatAction;
import org.telegram.telegrambots.meta.api.methods.send.SendMessage;
import org.telegram.telegrambots.meta.api.methods.send.SendSticker;
import org.telegram.telegrambots.meta.api.objects.InputFile;
import org.telegram.telegrambots.meta.api.objects.Message;
import org.telegram.telegrambots.meta.api.objects.Update;
import org.telegram.telegrambots.meta.exceptions.TelegramApiException;

import java.util.SplittableRandom;

import static com.telegram.folobot.Utils.printExeptionMessage;

@Component
@RequiredArgsConstructor
@Accessors(chain = true)
public class MessageService {

    @Setter
    private FoloBot foloBot;

    /**
     * Собрать объект {@link SendMessage}
     *
     * @param text   Текст сообщения
     * @param update {@link Update}
     * @return {@link SendMessage}
     */
    public SendMessage buildMessage(String text, Update update) {
        return SendMessage
                .builder()
                .parseMode(ParseMode.MARKDOWN)
                .chatId(Long.toString(update.getMessage().getChatId()))
                .text(text)
                .build();
    }

    /**
     * Собрать объект {@link SendMessage}
     *
     * @param text   Текст сообщения
     * @param update {@link Update}
     * @param reply  В ответ на сообщение
     * @return {@link SendMessage}
     */
    public SendMessage buildMessage(String text, Update update, boolean reply) {
        SendMessage sendMessage = buildMessage(text, update);
        if (reply) sendMessage.setReplyToMessageId(update.getMessage().getMessageId());
        return sendMessage;
    }

    /**
     * Отправить сообщение
     *
     * @param text   текст сообщения
     * @param chatid ID чата(пользователя)
     */
    public void sendMessage(String text, Long chatid) {
        try {
            foloBot.execute(SendMessage
                    .builder()
                    .parseMode(ParseMode.MARKDOWN)
                    .chatId(Long.toString(chatid))
                    .text(text)
                    .build());
        } catch (TelegramApiException e) {
            printExeptionMessage(e);
        }
    }

    /**
     * Отправить сообщение
     *
     * @param text   текст сообщения
     * @param update {@link Update}
     * @return {@link Message}
     */
    public Message sendMessage(String text, Update update) {
        try {
            return foloBot.execute(buildMessage(text, update));
        } catch (TelegramApiException e) {
            printExeptionMessage(e);
        }
        return null;
    }

    /**
     * Отправить сообщение буз реплая
     *
     * @param text   текст сообщения
     * @param update {@link Update}
     * @param reply  да/нет
     * @return {@link Message}
     */
    public Message sendMessage(String text, Update update, boolean reply) {
        if (!reply) {
            return sendMessage(text, update);
        } else {
            try {
                return foloBot.execute(buildMessage(text, update, reply));
            } catch (TelegramApiException e) {
                printExeptionMessage(e);
            }
        }
        return null;
    }

    private SendSticker buildSticker(InputFile stickerFile, Update update) {
        if (stickerFile != null) {
            return SendSticker
                    .builder()
                    .chatId(Long.toString(update.getMessage().getChatId()))
                    .sticker(stickerFile)
                    .replyToMessageId(update.getMessage().getMessageId())
                    .build();
        }
        return null;
    }

    /**
     * Отправить стикер
     *
     * @param stickerFile {@link InputFile}
     * @param update      {@link Update}
     */
    public void sendSticker(InputFile stickerFile, Update update) {
        if (stickerFile != null) {
            try {
                foloBot.execute(buildSticker(stickerFile, update));
            } catch (TelegramApiException e) {
                printExeptionMessage(e);
            }
        }
    }

    /**
     * Пересылка сообщений
     *
     * @param chatid Чат куда будет переслано сообщение
     * @param update {@link Update}
     */
    public void forwardMessage(Long chatid, Update update) {
        try {
            foloBot.execute(ForwardMessage
                    .builder()
                    .chatId(Long.toString(chatid))
                    .messageId(update.getMessage().getMessageId())
                    .fromChatId(Long.toString(update.getMessage().getChatId()))
                    .build());
        } catch (TelegramApiException e) {
            printExeptionMessage(e);
        }
    }

    public void forwardMessage(Long chatid, Message message) {
        if (message != null) {
            try {
                foloBot.execute(ForwardMessage
                        .builder()
                        .chatId(Long.toString(chatid))
                        .messageId(message.getMessageId())
                        .fromChatId(Long.toString(message.getChatId()))
                        .build());
            } catch (TelegramApiException e) {
                printExeptionMessage(e);
            }
        }
    }

    /**
     * Отправить статус "печатает"
     *
     * @param update {@link Update}
     */
    public void sendChatTyping(Update update) {
        try {
            foloBot.execute(SendChatAction
                    .builder()
                    .chatId(Long.toString(update.getMessage().getChatId()))
                    .action(String.valueOf(ActionType.TYPING))
                    .build());
        } catch (TelegramApiException e) {
            printExeptionMessage(e);
        }
    }

    /**
     * Получить случайный стикер (из гиф пака)
     *
     * @return {@link InputFile}
     */
    public InputFile getRandomSticker() {
        String[] stickers = {
                "CAACAgIAAxkBAAICCGKCCI-Ff-uqMZ-y4e0YmQEAAXp_RQAClxQAAnmaGEtOsbVbM13tniQE",
                "CAACAgIAAxkBAAPpYn7LsjgOH0OSJFBGx6WoIIKr_vcAAmQZAAJgRSBL_cLL_Nrl4OskBA",
                "CAACAgIAAxkBAAICCWKCCLoO6Itf6HSKKGedTPzbyeioAAJQFAACey0pSznSfTz0daK-JAQ",
                "CAACAgIAAxkBAAICCmKCCN_lePGRwqFYK4cPGBD4k_lpAAJcGQACmGshS9K8iR0VSuDVJAQ",
        };
        return new InputFile(stickers[new SplittableRandom().nextInt(stickers.length)]);
    }
}

File: src/main/kotlin/com/telegram/folobot/service/MessageService.kt
package com.telegram.folobot.service

import mu.KLogging
import org.springframework.stereotype.Component
import org.telegram.telegrambots.meta.api.methods.ActionType
import org.telegram.telegrambots.meta.api.methods.ForwardMessage
import org.telegram.telegrambots.meta.api.methods.ParseMode
import org.telegram.telegrambots.meta.api.methods.send.*
import org.telegram.telegrambots.meta.api.methods.updatingmessages.DeleteMessage
import org.telegram.telegrambots.meta.api.objects.InputFile
import org.telegram.telegrambots.meta.api.objects.Message
import org.telegram.telegrambots.meta.api.objects.Update
import org.telegram.telegrambots.meta.exceptions.TelegramApiException

@Component
class MessageService : KLogging() {
    lateinit var foloBot: FoloBot

    /**
     * Собрать объект [SendMessage]
     *
     * @param text   Текст сообщения
     * @param update [Update]
     * @return [SendMessage]
     */
    fun buildMessage(text: String, update: Update): SendMessage {
        return SendMessage
            .builder()
            .parseMode(ParseMode.MARKDOWN)
            .chatId(update.message.chatId.toString())
            .text(text)
            .build()
    }

    /**
     * Собрать объект [SendMessage]
     *
     * @param text   Текст сообщения
     * @param update [Update]
     * @param reply  В ответ на сообщение
     * @return [SendMessage]
     */
    fun buildMessage(text: String, update: Update, reply: Boolean): SendMessage {
        val sendMessage: SendMessage = buildMessage(text, update)
        if (reply) sendMessage.replyToMessageId = update.message.messageId
        return sendMessage
    }

    /**
     * Отправить сообщение
     *
     * @param text   текст сообщения
     * @param chatId ID чата(пользователя)
     */
    fun sendMessage(text: String, chatId: Long) {
        try {
            foloBot.execute(
                SendMessage
                    .builder()
                    .parseMode(ParseMode.MARKDOWN)
                    .chatId(chatId.toString())
                    .text(text)
                    .build()
            )
        } catch (e: TelegramApiException) {
            logger.error { e }
        }
    }

    /**
     * Отправить сообщение
     *
     * @param text   текст сообщения
     * @param update [Update]
     * @return [Message]
     */
    fun sendMessage(text: String, update: Update): Message? {
        try {
            return foloBot.execute(buildMessage(text, update))
        } catch (e: TelegramApiException) {
            logger.error { e }
        }
        return null
    }

    /**
     * Отправить сообщение буз реплая
     *
     * @param text   текст сообщения
     * @param update [Update]
     * @param reply  да/нет
     * @return [Message]
     */
    fun sendMessage(text: String, update: Update, reply: Boolean): Message? {
        if (!reply) {
            return sendMessage(text, update)
        } else {
            try {
                return foloBot.execute(buildMessage(text, update, reply))
            } catch (e: TelegramApiException) {
                logger.error { e }
            }
        }
        return null
    }

    private fun buildSticker(stickerId: String, update: Update): SendSticker? {
        return SendSticker
            .builder()
            .chatId(update.message.chatId.toString())
            .sticker(InputFile(stickerId))
            .replyToMessageId(update.message.messageId)
            .build()
    }

    /**
     * Отправить стикер
     *
     * @param stickerId [String]
     * @param update      [Update]
     */
    fun sendSticker(stickerId: String?, update: Update) {
        stickerId?.let {
            try {
                foloBot.execute(buildSticker(it, update))
            } catch (e: TelegramApiException) {
                logger.error { e }
            }
        }
    }

    /**
     * Пересылка сообщений
     *
     * @param chatId Чат куда будет переслано сообщение
     * @param update [Update]
     */
    fun forwardMessage(chatId: Long, update: Update) {
        try {
            foloBot.execute(
                ForwardMessage
                    .builder()
                    .chatId((chatId).toString())
                    .messageId(update.message.messageId)
                    .fromChatId(update.message.chatId.toString())
                    .build()
            )
        } catch (e: TelegramApiException) {
            logger.error { e }
        }
    }

    fun forwardMessage(chatId: Long, message: Message?) {
        message?.let {
            try {
                foloBot.execute(
                    ForwardMessage
                        .builder()
                        .chatId(chatId.toString())
                        .messageId(it.messageId)
                        .fromChatId(it.chatId.toString())
                        .build()
                )
            } catch (e: TelegramApiException) {
                logger.error { e }
            }
        }
    }

    /**
     * Отправить статус "печатает"
     *
     * @param update [Update]
     */
    fun sendChatTyping(update: Update) {
        try {
            foloBot.execute(
                SendChatAction
                    .builder()
                    .chatId(update.message.chatId.toString())
                    .action(ActionType.TYPING.toString())
                    .build()
            )
        } catch (e: TelegramApiException) {
            logger.error { e }
        }
    }

    /**
     * Отправить изображение
     *
     * @param photoId идентификатор изображения
     * @param text    текст сообщения
     * @param chatId  ID чата(пользователя)
     */
    fun sendPhoto(photoId: String, text: String, chatId: Long) {
        try {
            foloBot.execute(
                SendPhoto
                    .builder()
                    .parseMode(ParseMode.MARKDOWN)
                    .chatId(chatId.toString())
                    .photo(InputFile(photoId))
                    .caption(text)
                    .build()
            )
        } catch (e: TelegramApiException) {
            logger.error { e }
        }
    }

    /**
     * Отправить изображение
     *
     * @param photoId идентификатор изображения
     * @param text    текст сообщения
     * @param chatId  ID чата(пользователя)
     */
    fun sendPhotoFromResources(photoPath: String, text: String, chatId: Long) {
        try {
            foloBot.execute(
                SendPhoto
                    .builder()
                    .parseMode(ParseMode.MARKDOWN)
                    .chatId(chatId.toString())
                    .photo(
                        InputFile(
                            this::class.java.getResourceAsStream(photoPath),
                            photoPath.substringAfterLast("/")
                        )
                    )
                    .caption(text)
                    .build()
            )
        } catch (e: Exception) {
            logger.error { e }
        }
    }

    /**
     * Отправить аудио
     *
     * @param voiceId идентификатор audio
     * @param text    текст сообщения
     * @param chatId  ID чата(пользователя)
     */
    fun sendVoice(voiceId: String, text: String? = null, chatId: Long) {
        val voice = SendVoice
            .builder()
            .parseMode(ParseMode.MARKDOWN)
            .chatId(chatId.toString())
            .voice(InputFile(voiceId))
        text?.let { voice.caption(text) }
        try {
            foloBot.execute(voice.build())
        } catch (e: TelegramApiException) {
            logger.error { e }
        }
    }


    /**
     * Удалить сообщение
     *
     * @param update [Update]
     */
    fun deleteMessage(update: Update) {
        try {
            foloBot.execute(DeleteMessage(update.message.chatId.toString(), update.message.messageId))
        } catch (e: TelegramApiException) {
            logger.error { e }
        }
    }

    /**
     * Заместить сообщение
     *
     * @param update [Update]
     */
    fun substituteMessage(update: Update) {
        forwardMessage(update.message.chatId, update)
        deleteMessage(update)
    }

    /**
     * Получить случайный стикер (из гиф пака)
     *
     * @return [InputFile]
     */
    val randomSticker: String
        get() {
            return arrayOf(
                "CAACAgIAAxkBAAICCGKCCI-Ff-uqMZ-y4e0YmQEAAXp_RQAClxQAAnmaGEtOsbVbM13tniQE",
                "CAACAgIAAxkBAAPpYn7LsjgOH0OSJFBGx6WoIIKr_vcAAmQZAAJgRSBL_cLL_Nrl4OskBA",
                "CAACAgIAAxkBAAICCWKCCLoO6Itf6HSKKGedTPzbyeioAAJQFAACey0pSznSfTz0daK-JAQ",
                "CAACAgIAAxkBAAICCmKCCN_lePGRwqFYK4cPGBD4k_lpAAJcGQACmGshS9K8iR0VSuDVJAQ"
            ).random()
        }

    /**
     * Получить случайный войс
     *
     * @return [InputFile]
     */
    val randomVoice: String
        get() {
            return arrayOf(
                "AwACAgIAAx0CalJ4RAACBBVjsMmKi5mXdrptLdPr3l0iP1GgDwACPSAAAtdTaEmR_F2nU3-2SC0E",
                "AwACAgIAAx0CalJ4RAACBBZjsMmKzahNVLCHGsXr4rZI-kpr-AACQCAAAtdTaElCaBn1tNd33y0E",
                "AwACAgIAAx0CalJ4RAACBBdjsMmKJvi0kY6ZrQXcin5FjBZtTgACQSAAAtdTaEmfcn7TGEbnRS0E",
                "AwACAgIAAx0CalJ4RAACBBhjsMmKNJBkzTy9cDMcaAZsYhMVhwACRCAAAtdTaEmwShiRBaMuVS0E",
                "AwACAgIAAx0CalJ4RAACBBljsMmKfSlxH8HSoXn30vKeSZdZgAACRyAAAtdTaEnOjaACAklQ3S0E",
                "AwACAgIAAx0CalJ4RAACBBpjsMmKb-MRDMA84bs-QvS5-FQ83gACSSAAAtdTaEk15jd6A2irzi0E",
                "AwACAgIAAx0CalJ4RAACBBtjsMmKs3OkxcXJ4BGWy3rFIoU5ogACTyAAAtdTaElGaFICyutkdS0E",
                "AwACAgIAAx0CalJ4RAACBBxjsMmKjEZ1-nsrObYFz3E6OalKVAACUCAAAtdTaEmftSLquEa8GS0E",
                "AwACAgIAAx0CalJ4RAACBB1jsMmKlwPveslO3qG6hJGOWmpo2gACUSAAAtdTaEl8D6NwlexODy0E",
                "AwACAgIAAx0CalJ4RAACBB5jsMmKaVJ6KOLjI_todVKQ-9zyzgACViAAAtdTaEklG8wzFEN2Wi0E",
                "AwACAgIAAx0CalJ4RAACBB9jsMmK-rb94OCqEXrr8PHrAuKGWwACWiAAAtdTaEmXrdHjHqpcuS0E",
                "AwACAgIAAx0CalJ4RAACBCBjsMmKsIp1aANpMdd7qAyG9iU-FwACjSAAAtdTaEnvNbYGU1yo1C0E",
                "AwACAgIAAx0CalJ4RAACBCFjsMmK5VKXbBpYegU8KXObugLz8gACVCUAAjifcEkyH_coYlbNJS0E"
            ).random().also { logger.info { "Random voiceId is $it" } }

        }
}